### PR TITLE
Gold Pan Rework (Continuation of #3269)

### PR DESCRIPTION
**Pull request by JustAHuman-xD**
## Description
I'm making this PR to supersede #3269, this has all of svr333's changes + merge conflict fixes & other changes I realized could (and some should) be made when fixing the merge conflicts.

## Proposed changes
- Fix various typos in the Javadocs of  `ElectricGoldPan`
- Changes `GoldPan` & `NetherGoldPan` to use a `Set<Material>` for Inputs, rather than just one material
- Changes `GoldPan#getInputMaterial` to `GoldPan#getInputMaterials`
- Changes `GoldPan#getDisplayRecipes` to use a `ArrayList` instead of a `LinkedList`
- Adds `GoldPan#isValidInputMaterial`
- Adds `GoldPan#isValidInput`
- Adds missing `@Nonnull` annotations to various methods
- Adds Javadocs to `GoldPan#getGoldPanDrops`

## Related Issues (if applicable)
[Discord Suggestion #1614](https://discord.com/channels/565557184348422174/627048923122499584/887054754092580936)

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java [hunk 1]
```diff
@@ -40,31 +40,28 @@ public class ElectricGoldPan extends AContainer implements RecipeDisplayItem {
...
-     * overriden and allow the {@link ElectricGoldPan} to continue processing inputs
+     * overridden and allow the {@link ElectricGoldPan} to continue processing inputs
      * even if both output slots are occupied. Note this option will allow players
      * to force specific outputs from the {@link ElectricGoldPan} but can be
      * necessary when a server has disabled cargo networks.
      * 
-     * @return If output limits are overriden
+     * @return If output limits are overridden
      */
-    public boolean isOutputLimitOverriden() {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:62]: This is a public API, we would need a deprecation period for the old name

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 2]
```diff
@@ -59,14 +63,19 @@ public GoldPan(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeTyp
...
     /**
-     * This method returns the target {@link Material} for this {@link GoldPan}.
+     * This method returns the target {@link Material Materials} for this {@link GoldPan}.
      * 
-     * @return The {@link Material} this {@link GoldPan} can be used on
+     * @return The {@link Set} of {@link Material Materials} this {@link GoldPan} can be used on.
      */
-    public @Nonnull Material getInputMaterial() {
-        return Material.GRAVEL;
+    public @Nonnull Set<Material> getInputMaterials() {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:70]: Same here, public API, we can't change the signature without deprecation

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/NetherGoldPan.java [hunk 3]
```diff
@@ -19,18 +20,21 @@
...
+
     @ParametersAreNonnullByDefault
     public NetherGoldPan(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
         super(itemGroup, item, recipeType, recipe);
     }
 
     @Override
-    public @Nonnull Material getInputMaterial() {
-        return Material.SOUL_SAND;
+    public @Nonnull Set<Material> getInputMaterials() {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/NetherGoldPan.java:36]: Again here

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java (hunk 1 above)
Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:62]: should i have for removal set to true? or leave it as just Deprecated

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java [hunk 4]
```diff
@@ -80,30 +87,24 @@ public ItemStack getProgressBar() {
...
-                    return recipe;
-                }
+            if (goldPan.isValidInput(item)) {
+                output = goldPan.getRandomOutput();
+            } else if (netherGoldPan.isValidInput(item)) {
+                output = netherGoldPan.getRandomOutput();
+            }
 
+            if (output != null && output.getType() != Material.AIR && menu.fits(output, getOutputSlots())) {
+                MachineRecipe recipe = new MachineRecipe(4 / getSpeed(), new ItemStack[] { item }, new ItemStack[] { output });
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:105]: just asking about this tho, normal goldpan was 3/speed and nethergoldpan was 4/speed. is this change intentional?
should this also updates to lores and stuff on recipe (if any), or will it be automatically updated?

Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:105]: Oh good point, didn't catch that Ill change that rn

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/multiblocks/AutomatedPanningMachine.java [hunk 5]
```diff
@@ -62,38 +62,41 @@ public AutomatedPanningMachine(ItemGroup itemGroup, SlimefunItemStack item) {
     @Override
     public void onInteract(Player p, Block b) {
         ItemStack input = p.getInventory().getItemInMainHand();
+        Material material = input.getType();
+        ItemStack output;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/multiblocks/AutomatedPanningMachine.java:66]:
```suggestion
        ItemStack output = null;
```
for constitency we should set this to null

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

**Review by Sfiguz7 — APPROVED** [review]
LGTM

**Review by J3fftw1 — APPROVED** [review]

**Comment by Sfiguz7** [comment]
Compilation error :hmm:

**Comment by JustAHuman-xD** [comment]
> Compilation error :hmm:

It wasn't like that before you committed

**Comment by Sfiguz7** [comment]
Always blaming the elders
I guess the time has come to revert that

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/multiblocks/AutomatedPanningMachine.java [hunk 6]
```diff
@@ -62,38 +62,41 @@ public AutomatedPanningMachine(ItemGroup itemGroup, SlimefunItemStack item) {
     @Override
     public void onInteract(Player p, Block b) {
         ItemStack input = p.getInventory().getItemInMainHand();
+        Material material = input.getType();
+        ItemStack output = null;
```
Comment (Sfiguz7) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/multiblocks/AutomatedPanningMachine.java:66]:
```suggestion
        ItemStack output;
```

**Review by Sfiguz7 — CHANGES_REQUESTED** [review]
Noone saw this

**Comment by JustAHuman-xD** [comment]
Xd

**Review by Sfiguz7 — APPROVED** [review]
I am dumb, there you go

**Comment by JustAHuman-xD** [comment]
> I am dumb, there you go

You are not dumb

**Comment by JustAHuman-xD** [comment]
Also I'll send a message once I've tested everything myself. Don't merge into then or until someone else tests it.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java [hunk 7]
```diff
@@ -80,30 +87,26 @@ public ItemStack getProgressBar() {
...
-            } else if (SlimefunUtils.isItemSimilar(item, soulSand, true, false)) {
-                ItemStack output = netherGoldPan.getRandomOutput();
-                MachineRecipe recipe = new MachineRecipe(4 / getSpeed(), new ItemStack[] { soulSand }, new ItemStack[] { output });
-
-                if (output.getType() != Material.AIR && menu.fits(output, getOutputSlots())) {
-                    menu.consumeItem(slot);
-                    return recipe;
-                }
-
+            if (output != null && output.getType() != Material.AIR && menu.fits(output, getOutputSlots())) {
```
[the remark this answers is not in the document]
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:107]: i think you are right here

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 8]
```diff
@@ -160,16 +181,46 @@ public void updateRandomizer() {
...
+    /**
+     * This returns whether the {@link GoldPan} accepts the {@link ItemStack} as an input
+     *
+     * @param itemStack
+     *            The {@link ItemStack} to check
+     *
+     * @return If the {@link ItemStack} is valid
+     */
+    public boolean isValidInput(ItemStack itemStack) {
+        Material material = itemStack.getType();
```
Comment (Sfiguz7) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:209]: This is the issue, if itemStack == null this fails and that's what triggers the error. Same would happen with the nether gold pan

**Review by Sfiguz7 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sfiguz7) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:107]: The error is caused by something very similar but 10 lines above, I commented below requesting changes. Will resolve this due to that

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java [hunk 9]
```diff
@@ -40,31 +40,38 @@ public class ElectricGoldPan extends AContainer implements RecipeDisplayItem {
...
         addItemSetting(overrideOutputLimit);
     }
 
     /**
-     * This returns whether the {@link ElectricGoldPan} will stop proccessing inputs
+     * @deprecated
+     *
+     * @see ElectricGoldPan#isOutputLimitOverridden()
+     */
+    @Deprecated
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:54]: We should really add deprecated since but thats also a pain to do.
Since we only care about RC and there is an RC releasing this week and i doubt this will be in it can we add @Deprecated since: "RC 36"

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 10]
```diff
@@ -59,14 +63,31 @@ public GoldPan(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeTyp
...
-     * 
-     * @return The {@link Material} this {@link GoldPan} can be used on
+     * @deprecated
+     *
+     * This method returns the old {@link Material} that this {@link GoldPan} accepted
+     *
+     * @return The old {@link Material} this {@link GoldPan} could be used on
      */
-    public @Nonnull Material getInputMaterial() {
+    @Deprecated
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:72]: same here

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 11]
```diff
@@ -59,14 +63,31 @@ public GoldPan(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeTyp
...
 
     /**
-     * This method returns the target {@link Material} for this {@link GoldPan}.
-     * 
-     * @return The {@link Material} this {@link GoldPan} can be used on
+     * @deprecated
+     *
+     * This method returns the old {@link Material} that this {@link GoldPan} accepted
+     *
+     * @return The old {@link Material} this {@link GoldPan} could be used on
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:70]: i get what you mean but i read this literal you dont return the old material

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java (hunk 9 above)
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:54]: example: https://github.com/Slimefun/Slimefun4/blob/75a781d3d651b6db7fe6697ecc48bd28e486cd49/src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/LongFallBoots.java#L28

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java (hunk 11 above)
Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:70]: I don't think I follow

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java (hunk 9 above)
Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:54]: > example: 
> https://github.com/Slimefun/Slimefun4/blob/75a781d3d651b6db7fe6697ecc48bd28e486cd49/src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/LongFallBoots.java#L28

Should I add exactly that?

**Comment by variananora** [comment]
Just tested it and it is working correctly.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java (hunk 11 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:70]: Remove the old words. You are implying that those are old materials even though they arent old.

Comment (Sfiguz7) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java]: Maybe "legacy" is a better word? It's a method that I marked for removal anyway on the GH project (or that I should mark for removal)

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java]: I mean it’s not legacy either to method stays the same and the material isn’t legacy docs doesn’t need to be changed when it’s marked as deprecated

Comment (Sfiguz7) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java]: Aaah I see what you mean. Yeah ok makes sense to me now

Comment (JustAHuman-xD) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java]: Got it, I’ll change the wording in a bit then, can’t atm

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java [hunk 12]
```diff
@@ -40,31 +40,38 @@ public class ElectricGoldPan extends AContainer implements RecipeDisplayItem {
...
     public ElectricGoldPan(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
         super(itemGroup, item, recipeType, recipe);
         addItemSetting(overrideOutputLimit);
     }
 
     /**
-     * This returns whether the {@link ElectricGoldPan} will stop proccessing inputs
+     * @deprecated since RC-36
+     *
+     * @see ElectricGoldPan#isOutputLimitOverridden()
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/ElectricGoldPan.java:52]:
```suggestion
     * Use {@link ElectricGoldPan#isOutputLimitOverridden()} instead.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 13]
```diff
@@ -59,14 +63,29 @@ public GoldPan(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeTyp
     }
 
     /**
-     * This method returns the target {@link Material} for this {@link GoldPan}.
-     * 
-     * @return The {@link Material} this {@link GoldPan} can be used on
+     * @deprecated since RC-36
+     *
+     * @see GoldPan#getInputMaterials()
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:68]:
```suggestion
     * Use {@link GoldPan#getInputMaterials()} instead.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 14]
```diff
@@ -160,16 +179,49 @@ public void updateRandomizer() {
...
 
+    /**
+     * This returns whether the {@link GoldPan} accepts the {@link ItemStack} as an input
+     *
+     * @param itemStack
+     *            The {@link ItemStack} to check
+     *
+     * @return If the {@link ItemStack} is valid
+     */
+    public boolean isValidInput(ItemStack itemStack) {
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:206]:
```suggestion
    public boolean isValidInput(@Nullable ItemStack itemStack) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java [hunk 15]
```diff
@@ -160,16 +179,49 @@ public void updateRandomizer() {
...
+
+    /**
+     * This returns whether the {@link GoldPan} accepts the {@link Material} as an input
+     *
+     * @param material
+     *            The {@link Material} to check
+     *
+     * @return If the {@link Material} is valid
+     */
+    public boolean isValidInputMaterial(Material material) {
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/GoldPan.java:223]:
```suggestion
    public boolean isValidInputMaterial(@Nonnull Material material) {
```

Goign to take a stab this is nonnull, though check and if from context it can be nullable, annotate it as such instead.

**Review by Sefiraat — CHANGES_REQUESTED** [review]

**Review by J3fftw1 — APPROVED** [review]
LGTM

**Review by Sfiguz7 — APPROVED** [review]

**Comment by Sfiguz7** [comment]
@Sefiraat since we did not mark your review as stale could you please double check this? It's good to be merged otherwise
